torch/csrc/autograd/input_metadata.h

```diff
@@ -42,6 +52,13 @@ struct InputMetadata {
 private:
   const at::Type* type_ = nullptr;
   at::DimVector shape_;
+  // NB: This looks like a "device" but it actually is an assignment to
+  // an autograd thread; e.g., both CUDA device 1 and XLA device 1 live
+  // on the same thread.  We should change this to *actually* be a Device
+  // but we have to fix all of the sites to adjust this way.  (Why is
+  // it this way?  In the old days we only had CUDA devices, and so
+  // we could just say -1 was CPU and everything else was CUDA.  This clearly
+  // doesn't hold anymore.)
```

Comment:
This feels wrong. We should really use `Device` here, and the decision process that the engine uses to choose the thread shouldn't leak into here. Can't we just change those checks to use the real device, and contain the other logic? It doesn't seem to be _that_ hard. It's just pushing the call to `assign_tensor_to_autograd_thread` to a later point.

Reply:
IMHO there should be to pinning of thread->device at thread boot time.
We should spawn an adequately sized pool, and let them fetch "work". The "work" will carry a  device, and the proper device guard should be set while that thread is processing that work.

Reply:
@apaszke I attempted to do this months ago and there are a few tricky interaction points. I'd much rather do this refactor in a separate patch.

Reply:
@dlibenzi it's actually not always beneficial to have thread pools there, hence this strategy.

@ezyang ok it doesn't have to be the full refactor but the comment above this one should be fixed before this lands. As is this patch is uber bug prone.

Reply:
@apaszke For thread pool, I did not mean it in any dynamic way. Today PT is starting one thread per device, and binds it to a device. So for example if one have 16 devices and 4 cores, given that the work is CPU bound, we can have thrashing.
If PT instead starts min(num_devices, num_cores) threads, and does not bind the threads to devices at boot time, but instead let the device come in as part of the "work" the threads will pull from the queue, it seems the logic becomes simpler by removing the it-this-then-that logic in the thread init function, where PT creates the device guards.
The thread working loop can be like:
```
while (true) {
  auto work = thread_queue.get();
  auto device_guard = CreateDeviceGuard(work->device);
  process(work);
}
```
And the thread init function will lose a bunch of code.